zeppelin-interpreter/src/main/java/org/apache/zeppelin/cluster/ClusterManager.java

```diff
@@ -0,0 +1,527 @@
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.DELETE_OPERATION;
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.PUT_OPERATION;
+import static org.apache.zeppelin.cluster.meta.ClusterMetaOperation.GET_OPERATION;
+
+/**
+ * The base class for cluster management, including the following implementations
+ * 1. RaftClient as the raft client
+ * 2. Threading to provide retry after cluster metadata submission failure
+ * 3. Cluster monitoring
+ */
+public abstract class ClusterManager {
+  private static Logger logger = LoggerFactory.getLogger(ClusterManager.class);
+
+  public final ZeppelinConfiguration zconf = ZeppelinConfiguration.create();
+
+  protected Collection<Node> clusterNodes = new ArrayList<>();
+
+  // raft
+  protected static String ZEPL_CLUSTER_ID = "ZEPL-CLUSTER";
+  protected static String ZEPL_CLIENT_ID = "ZEPL-CLIENT";
+
+  protected int raftServerPort = 0;
+
+  protected RaftClient raftClient = null;
+  protected SessionClient raftSessionClient = null;
+  protected Map<MemberId, Address> raftAddressMap = new ConcurrentHashMap<>();
+  protected LocalRaftProtocolFactory protocolFactory
+      = new LocalRaftProtocolFactory(protocolSerializer);
+  protected List<MessagingService> messagingServices = new ArrayList<>();
+  protected Collection<MemberId> clusterMemberIds = new ArrayList<MemberId>();
+
+  protected AtomicBoolean running = new AtomicBoolean(true);
+
+  // Write data through the queue to prevent failure due to network exceptions
+  private ConcurrentLinkedQueue<ClusterMetaEntity> clusterMetaQueue
+      = new ConcurrentLinkedQueue<>();
+
+  // zeppelin server host & port
+  protected String zeplServerHost = "";
+
+  // atomic dependent netty package path
+  public static String ATOMIX_NETTY_LIB_PATH = "/lib/atomix-netty";
+  public static String INTERPRETER_LIB_PATH = "/lib/interpreter";
+
+  public ClusterManager() {
+    try {
+      zeplServerHost = RemoteInterpreterUtils.findAvailableHostAddress();
+      String clusterAddr = zconf.getClusterAddress();
+      if (!StringUtils.isEmpty(clusterAddr)) {
+        String cluster[] = clusterAddr.split(",");
+
+        for (int i = 0; i < cluster.length; i++) {
+          String[] parts = cluster[i].split(":");
+          String clusterHost = parts[0];
+          int clusterPort = Integer.valueOf(parts[1]);
+          if (zeplServerHost.equalsIgnoreCase(clusterHost)) {
+            raftServerPort = clusterPort;
+          }
+
+          Node node = Node.builder().withId(cluster[i])
+              .withAddress(Address.from(clusterHost, clusterPort)).build();
+          clusterNodes.add(node);
+          raftAddressMap.put(MemberId.from(cluster[i]), Address.from(clusterHost, clusterPort));
+          clusterMemberIds.add(MemberId.from(cluster[i]));
+        }
+      }
+    } catch (UnknownHostException e) {
+      e.printStackTrace();
+    } catch (SocketException e) {
+      e.printStackTrace();
+      return;
```

Comment:
return is not necessary for constructor.

Reply:
done